pkg/clustermesh/remote_cluster.go

```diff
@@ -107,45 +115,58 @@ func (rc *remoteCluster) getLogger() *logrus.Entry {
 	})
 }
 
+// releaseOldConnection releases the etcd connection to a remote cluster
 func (rc *remoteCluster) releaseOldConnection() {
-	if rc.ipCacheWatcher != nil {
-		rc.ipCacheWatcher.Close()
-		rc.ipCacheWatcher = nil
-	}
+	rc.mutex.Lock()
+	ipCacheWatcher := rc.ipCacheWatcher
+	rc.ipCacheWatcher = nil
 
-	if rc.remoteNodes != nil {
-		rc.remoteNodes.Close(context.TODO())
-		rc.remoteNodes = nil
-	}
-	if rc.remoteIdentityCache != nil {
-		rc.remoteIdentityCache.Close()
-		rc.remoteIdentityCache = nil
-	}
-	if rc.remoteServices != nil {
-		rc.remoteServices.Close(context.TODO())
-		rc.remoteServices = nil
-	}
-	if rc.backend != nil {
-		rc.backend.Close()
-		rc.backend = nil
-	}
+	remoteNodes := rc.remoteNodes
+	rc.remoteNodes = nil
+
+	remoteIdentityCache := rc.remoteIdentityCache
+	rc.remoteIdentityCache = nil
+
+	remoteServices := rc.remoteServices
+	rc.remoteServices = nil
+
+	backend := rc.backend
+	rc.backend = nil
+	rc.mutex.Unlock()
+
+	// Release resources asynchroneously in the background. Many of these
+	// operations may time out if the connection was closed due to an error
+	// condition.
```

Comment:
The implementation isn't really changing here but I note that the contexts have no deadline below, and was wondering whether `remoteNodes.Close()` or `remoteServices.Close()` could block indefinitely?

(Could be a follow-up)

Reply:
Note that if the answer is "yes" then we need to deal with the other `context.TODO()` in the controller failure cases below as well...

Reply:
Related: https://github.com/cilium/cilium/pull/12526

Reply:
I had looked at that and while it should have a timeout, it doesn't matter that much for clustermesh as the blocking operation is the deletion of local keys. The SharedStore is only used to listen so no local keys have to be deleted when the store is closed. We can fix it outside of this PR.